design/one-pager-resource-connectivity-mvp.md

```diff
@@ -0,0 +1,653 @@
+configured such that they can communicate with each other at a network level.
+This configuration process varies from cloud provider to cloud provider and from
+external resource to external resource. Frequently the operator of the external
+resources is required to ensure they're both in the same cloud region, and/or
+the same VPC network, subnetwork, security group, etc in order to communicate.
+Certain settings may also need to be enabled on one or both external resources.
+This highlights two shortcomings in Crossplane:
+
+* Managed resources are not consistently 'high fidelity', in that they don't
+  expose all of the settings their underlying external resource's API exposes.
+  Crossplane does not consistently expose the settings necessary to configure
+  connectivity.
+* External resources often depend on other external resources, particularly
+  network constructs like VPC networks, in order to configure connectivity.
+  Crossplane does not model these external resources as managed resources,
+  requiring the cloud administrator to create and manage them outside of
+  Crossplane.
+
+This document will walk through an example scenario for each supported cloud
+provider in which an application operator wishes to deploy [Wordpress] to a
+`KubernetesCluster`, backed by a `MySQLInstance`, highlighting the minimum
+changes necessary for Crossplane to support this.
+
+### Low Fidelity Managed Resources
+
+The below examples illustrate the fidelity of two Crossplane managed resources
+in relation to their equivalent API objects. In each case a comprehensive YAML
+document representing every supported field of the Crossplane managed resource
+is compared to a hypothetical YAML document resulting from directly translating
+the cloud provider API. Note that these examples are primarily intended to
+illustrate the at times significant difference between the configuration a cloud
+provider API supports, and the configuration Crossplane exposes. _Complete_
+fidelity in Crossplane managed resources is not necessary to enable connectivity
+between managed resources; it would be sufficient in the context of this design
+to expose only any missing connectivity-related fields.
+
+### CloudSQL
+
+First, an exhaustive example of the settings Crossplane's `CloudSQLInstance`
+currently supports. The `CloudSQLInstance` resource claim controller supports
+reading all of the below settings from a `ResourceClass` at dynamic provisioning
+time.
+
+```yaml
+---
+apiVersion: database.gcp.crossplane.io/v1alpha1
+kind: CloudSQLInstance
+metadata:
+  namespace: crossplane-system
+  name: example
+spec:
+  nameFormat: mycoolname
+  tier: db-n1-standard-1
+  authorizedNetworks:
+  - 73.98.0.0/16
+  databaseVersion: MYSQL_5_6
+  region: us-central1
+  storageType: PD_SSD
+  storageGB: 50
+  labels:
+    cool: very
+```
+
+By comparison, a direct translation of the [CloudSQL external resource]'s
+writable API object fields to a Kubernetes YAML specification would be as
```

Comment:
It is not as straight forward as this sentence seems to imply. 
In this specific case `CloudSQL` the instance can be created using 3 different instance types (`InstanceType`) and based on the type, some properties fields are "writable" while others "read-only", and "some others" are n/a. 

For example, creating `InstanceType`: `CLOUD_SQL_INSTANCE`, we don't need to provide `masterInstanceName`, wheras, we must provide it for `READ_REPLICA_INSTANCE`. Replicas types themselves could be of `FAILOVER` or `READ` types, which also have some overlapping and some disticnt set of properties.

Reply:
My intention here is to suggest that Crossplane's `CloudSQLInstanceSpec` be as close as reasonably possible to the fields exposed by the GCP [`DatabaseInstance`](https://godoc.org/google.golang.org/api/sqladmin/v1beta4#DatabaseInstance) excluding the "output only" fields (which are good candidates to expose in `CloudSQLInstanceStatus` instead). I take your point that the value of one field may influence which other fields are writable. My opinion is that at the managed resource level we should faithfully expose all the knobs to the managed resource author and let them figure it out.

Consider how the CloudSQL SDK works; the author is expected to understand the API documentation and know that if they set `InstanceType: CLOUD_SQL_INSTANCE` then they should not set `MasterInstanceName`. My guess is that `MasterInstanceName` is simply ignored if `InstanceType: CLOUD_SQL_INSTANCE` is set, but for the sake of this example let's assume the API returns an error when a `DatabaseInstance` is submitted with `InstanceType: CLOUD_SQL_INSTANCE` and `MasterInstanceName: mycoolmaster`.

I propose that Crossplane, at the managed resource level, does nothing to stop the managed resource author writing this nonsensical managed resource, just as the CloudSQL Go SDK does nothing to prevent the author submitting this nonsensical `DatabaseInstance`. I propose that Crossplane faithfully translate the `CloudSQLInstanceSpec` into a `DatabaseInstance`, try create it, and return the inevitable error to the author by way of a `ReconcileError` `ConditionedStatus`.

This makes the mental model for managing Crossplane resources as close as possible to the mental model of using the underlying cloud provider API (which is typically well documented, and has Googlable error strings). It also keeps the path a little clearer for generating `Reconciler` code from a cloud provider API specification.